packages/server/src/utils/workflow.ts

```diff
@@ -1,106 +1,72 @@
 import { LlamaIndexAdapter, StreamData, type JSONValue } from "ai";
 import type {
-  AgentInputData,
   ChatResponseChunk,
   EngineResponse,
   Metadata,
   NodeWithScore,
-  WorkflowEvent,
+  WorkflowEventData,
+  WorkflowStream,
 } from "llamaindex";
 import {
-  AgentStream,
-  AgentToolCall,
-  AgentToolCallResult,
+  agentStreamEvent,
+  agentToolCallEvent,
+  agentToolCallResultEvent,
   AgentWorkflow,
   LLamaCloudFileService,
-  StopEvent,
-  Workflow,
-  type AgentWorkflowContext,
+  stopAgentEvent,
 } from "llamaindex";
 import { ReadableStream } from "stream/web";
 import {
-  SourceEvent,
+  sourceEvent,
   toAgentRunEvent,
   toSourceEvent,
   type SourceEventNode,
 } from "../events";
-import type { ServerWorkflow } from "../types";
+import {
+  workflowInputEvent,
+  type ServerWorkflow,
+  type WorkflowInput,
+} from "../types";
 import { downloadFile } from "./file";
-import { sendSuggestedQuestionsEvent } from "./suggestion";
 
 export async function runWorkflow(
   workflow: ServerWorkflow,
-  agentInput: AgentInputData,
+  input: WorkflowInput,
 ) {
+  const dataStream = new StreamData();
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  let workflowStream: AsyncIterable<WorkflowEventData<any>> | WorkflowStream;
   if (workflow instanceof AgentWorkflow) {
-    return runAgentWorkflow(workflow, agentInput);
+    workflowStream = workflow.runStream(input.userInput, {
+      chatHistory: input.chatHistory,
+    });
+  } else {
+    // TODO: Refactor this using stream API from llamaindex once it's ready
+    const { stream, sendEvent } = workflow.createContext();
+    sendEvent(
+      workflowInputEvent.with({
+        userInput: input.userInput,
+        chatHistory: input.chatHistory,
+      }),
+    );
+    workflowStream = stream;
   }
-  return runCustomWorkflow(workflow, agentInput);
-}
-
-async function runAgentWorkflow(
-  workflow: AgentWorkflow,
-  agentInput: AgentInputData,
-) {
-  const { userInput = "", chatHistory = [] } = agentInput;
-  const context = workflow.run(userInput, { chatHistory });
-
-  const dataStream = new StreamData();
 
   const stream = new ReadableStream<EngineResponse>({
     async pull(controller) {
       try {
-        for await (const event of context) {
-          if (event instanceof AgentStream) {
-            // for agent workflow, get the delta from AgentStream event and enqueue it
-            const delta = event.data.delta;
-            if (delta) {
-              controller.enqueue({ delta } as EngineResponse);
-            }
-          } else {
-            appendEventDataToAnnotations(dataStream, event);
+        for await (const event of workflowStream) {
```

Comment:
you might be able to simplify this be using a `TransformStream` on workflowStream that transforms it to a vercel compatible data stream

have a look at the LlamaIndexAdapter code: https://github.com/vercel/ai/blob/845080d80b8538bb9c7e527d2213acb5f33ac9c2/packages/ai/streams/llamaindex-adapter.ts

Reply:
calling `for await` inside a `pull` is not a good idea, as it consumes all events of workflowStream whereas the outside reader just want to pull one item